proxy/nasshp/nassh.go

```diff
@@ -103,8 +118,135 @@ type options struct {
+		"How often to check if there are sessions to be expired. Zero disables the check.")
+	set.IntVar(&ep.RuthlessThreshold, prefix+"expire-ruthlessly-after", ep.RuthlessThreshold,
+		"If more than this number of sessions are orphaned, the oldest sessions will be "+
+			"expired until fewer than this limit sessions are left orphaned.")
+	set.IntVar(&ep.OrphanThreshold, prefix+"expire-long-after", ep.OrphanThreshold,
+		"If more than this number of sessions are orphaned, the oldest sessions that "+
+			"have been orphaned for longer than expire-long-time will be expired.")
+	set.DurationVar(&ep.OrphanLimit, prefix+"expire-long-time", ep.OrphanLimit,
+		"Sessions orphaned for less than this time will not be expired unless the "+
+			"ruthless limit is exceeded")
+}
+
+func (ep *ExpirationPolicy) Expire(ctx context.Context, clock utils.Clock, sessions *sessions, counters *ExpireCounters) {
+	var start time.Time
+	for {
+		counters.ExpireRuns.Increment()
+		if !start.IsZero() {
+			end := clock.Now()
+			counters.ExpireDuration.Add(uint64(end.Sub(start)))
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-clock.After(ep.Every):
+		}
+
+		start = clock.Now()
+
+		total := sessions.Created.Get() - sessions.Deleted.Get()
+		if total < uint64(ep.OrphanThreshold) {
+			continue
+		}
+
+		counters.ExpireAboveOrphanThresholdRuns.Increment()
+		counters.ExpireAboveOrphanThresholdTotal.Add(total)
+
+		// Save the time BEFORE we start scanning sessions (so we can
+		// exclude all sessions (un)paused after the scan), and find
+		// all sessions paused.
+		now := clock.Now().UnixNano()
+		orphaned := []*readWriter{}
+		sessions.Iterate(func(rw *readWriter) bool {
+			pausedat := rw.browser.paused.Nano()
+			if pausedat <= 0 || pausedat > now {
+				return true
+			}
+
+			orphaned = append(orphaned, rw)
+			return true
+		})
+
+		counters.ExpireAboveOrphanThresholdFound.Add(uint64(len(orphaned)))
+		if len(orphaned) < ep.OrphanThreshold {
+			continue
+		}
+
+		sort.Slice(orphaned, func(i, j int) bool {
+			return orphaned[i].browser.paused.Nano() < orphaned[j].browser.paused.Nano()
+		})
+
+		maxorphan := int64(ep.OrphanLimit * time.Nanosecond)
+		for cursor, left := 0, len(orphaned); cursor < len(orphaned) && left > ep.OrphanThreshold; cursor, left = cursor+1, left-1 {
+			rw := orphaned[cursor]
+
+			// This can happen if the session was unpaused in
+			// between scanning the map and processing the list of
+			// sessions. No matter what, there's a short race here.
+			pausedon := rw.browser.paused.Nano()
+			pausedfor := now - pausedon
+			if pausedon <= 0 || pausedfor <= 0 {
+				counters.ExpireRaced.Increment()
+				continue
+			}
+
+			if left <= ep.RuthlessThreshold {
+				if pausedfor <= maxorphan {
+					break
+				}
+				counters.ExpireOrphanClosed.Increment()
+			} else {
+				counters.ExpireRuthlessClosed.Increment()
+			}
+
+			counters.ExpireLifetimeTotal.Add(uint64(pausedfor / 1e9))
```

Comment:
`uint64(pausedFor) / uint64(time.Second)` might be clearer?

Reply:
good suggestion, done.